docs_src/design/adr/core/0021-Device-Profile-Changes.md

```diff
@@ -0,0 +1,64 @@
+- A device profile cannot be removed when it is associated to a device or provision watcher.
+- A device profile can be removed or modified even when associated to an event or reading unless configuration options (see New Configuration Settings below) are set prohibiting removal or modification when associated to an event/reading
+    - the rationale behind this is that events and readings are generally considered short lived (ephemeral) objects and already contain the necessary device profile information that are needed by the system during their short life without having to refer to the device profile
+    - see note below in Consequences that a new Units property must be added to the Reading object in order to support this rule and the need for all relevant profile data to be in the event/reading.
+
+### Ancillary Rules associated to Device Profiles
+- Identifying or “key” fields for device profiles, device resources, etc. cannot be modified and can never be null.
+- A device profile can begin life “empty” - meaning that it has no device resources or device commands.
+
+### New APIs
+
+The following APIs would be added to the metadata REST service in order to meet the design specified above.
+
+- Add Profile General Property PATCH API (allow to modify profile properties except name and id)
+- Add Profile Device Resource POST API
+- Add Profile Device Resource PATCH API (allow to modify Description and IsHidden only)
+- Add Profile Device Resource DELETE API (allow as described above)
+- Add Profile Device Command POST API
+- Add Profile Device Command PATCH API (allow as described above)
+- Add Profile Device Command DELETE API (allow as described above)
+
+### New Configuration Settings
+
+Some adopters may not view event/reading data as ephemeral or short lived.  These adopters may choose not to allow device profiles to be modified or removed when associated to an event or reading.
+For this reason, two new configuration options will be added to metadata configuration that are used to reject modifications or deletions.
+
+- AllowEventAssociatedDeviceProfileModifications (set to true by default)
+- AllowEventAssociatedDeviceProfileDeletes (set to true by default)
+
+When either of these config settings are set to false, metadata would accordingly reject changes to or removal of profiles when the profile is associated to any event/reading.
+
+## Consequences/Considerations
+In order to allow device profiles to be updated or removed even when associated to an EdgeX event/reading, a new property needs to be added to the reading object.
+
+- Readings will now contain a “Units” (string) property.  This property will indicate the units of measure for the Value in the Reading and will be populated based on the Units for the device resource.
```

Comment:
Some users complained about the size of Event/Readings.  May we only add Readings and other relevant information when any of the New Configuration Settings is false?
If the associations are both true, Units is redundant in Readings

Reply:
We're talking about pretty short strings (e.g. "Units": "UCOM:Cel"), however if we really think this is a concern, we could add an additional setting (e.g. `ReadingUnits`) that defaults to `true`.

Reply:
Those user would generate more than one thousand Readings each second, so that should be considered

Reply:
Add new config setting per above.  I would assume this could be Writable section of the device services area (and SDKs too)?